zk-token-sdk/src/encryption/grouped_elgamal.rs

```diff
@@ -0,0 +1,287 @@
+    /// decryption handle at a specified index.
+    ///
+    /// The output of this function is of type `DiscreteLog`. To recover the originally encrypted
+    /// amount, use `DiscreteLog::decode`.
+    fn decrypt(
+        grouped_ciphertext: &GroupedElGamalCiphertext<N>,
+        secret: &ElGamalSecretKey,
+        index: usize,
+    ) -> Result<DiscreteLog, GroupedElGamalError> {
+        Self::to_elgamal_ciphertext(grouped_ciphertext, index)
+            .map(|ciphertext| ciphertext.decrypt(secret))
+    }
+
+    /// Decrypts a grouped ElGamal ciphertext to a number that is interpretted as a positive 32-bit
+    /// number (but still of type `u64`).
+    ///
+    /// If the originally encrypted amount is not a positive 32-bit number, then the function
+    /// returns `None`.
+    fn decrypt_u32(
+        grouped_ciphertext: &GroupedElGamalCiphertext<N>,
+        secret: &ElGamalSecretKey,
+        index: usize,
+    ) -> Result<Option<u64>, GroupedElGamalError> {
+        Self::to_elgamal_ciphertext(grouped_ciphertext, index)
+            .map(|ciphertext| ciphertext.decrypt_u32(secret))
+    }
+}
+
+/// A grouped ElGamal ciphertext.
+///
+/// The type is defined with a generic constant constant parameter that specifies the number of
+/// decryption handles that the ciphertext holds.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct GroupedElGamalCiphertext<const N: usize> {
+    pub commitment: PedersenCommitment,
+    pub handles: [DecryptHandle; N],
+}
+
+impl<const N: usize> GroupedElGamalCiphertext<N> {
+    /// Decrypts the grouped ElGamal ciphertext using an ElGamal secret key pertaining to a
+    /// specified index.
+    ///
+    /// The output of this function is of type `DiscreteLog`. To recover the originally encrypted
+    /// amount, use `DiscreteLog::decode`.
+    pub fn decrypt(
+        &self,
+        secret: &ElGamalSecretKey,
+        index: usize,
+    ) -> Result<DiscreteLog, GroupedElGamalError> {
+        GroupedElGamal::decrypt(self, secret, index)
+    }
+
+    /// Decrypts the grouped ElGamal ciphertext to a number that is interpretted as a positive 32-bit
+    /// number (but still of type `u64`).
+    ///
+    /// If the originally encrypted amount is not a positive 32-bit number, then the function
+    /// returns `None`.
+    pub fn decrypt_u32(
+        &self,
+        secret: &ElGamalSecretKey,
+        index: usize,
+    ) -> Result<Option<u64>, GroupedElGamalError> {
+        GroupedElGamal::decrypt_u32(self, secret, index)
+    }
+
+    /// The expected length of a serialized grouped ElGamal ciphertext.
+    ///
+    /// A grouped ElGamal ciphertext consists of a Pedersen commitment and an array of decryption
+    /// handles. The commitment and decryption handles are a single Curve25519 group element each
+    /// that is serialized as bytes of length 32.
+    fn expected_byte_length() -> usize {
+        N.saturating_add(1).saturating_mul(32)
```

Comment:
It's not immediately obvious why there is a `.saturating_add(1)`.
Generally, there are a lot of magic numbers in this crate. I don't mean to bog things down, but what do you think about a PR constructing constants for a bunch of them?
So this line would become something like:
```
PEDERSEN_COMMITMENT_LEN + N.saturating_mul(CIPHERTEXT_LEN)
```

Reply:
Ah yes you are right. The ciphertext consists of a commitment (32 bytes) and N decryption handles (32 bytes), so the expected length is (N+1)*32. I used `saturating_add` here, but it does seem like using `checked_add` and then unwrapping makes more sense here since N should never be big enough so overflow `usize`.

Reply:
I don't mean to bog things down, but what do you think about a PR constructing constants for a bunch of them? So this line would become something like:
> 
> ```
> PEDERSEN_COMMITMENT_LEN + N.saturating_mul(CIPHERTEXT_LEN)
> ```
Yes, you are completely right. I have been meaning to do that at some point... I will make a PR once the last bits of the proof program is finished.